cmake/ccache.cmake

```diff
@@ -0,0 +1,93 @@
+#[[
+Copyright 2023 Antony Peacock
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+]]
+
+include_guard(GLOBAL)
+
+
+#[=======================================================================[.rst:
+enable_ccacche
+------------------
+
+Overview
+^^^^^^^^
+
+Enable use of Ccache for compiler caching during the build process.
+
+.. code-block:: cmake
+
+  enable_ccacche(
+      [QUIET]
+  )
+   -- Searches for the ccache executable, and if found enables it for supported
+   language with specified options.
+
+  ``QUIET`` Silent output from this method.
+
+#]=======================================================================]
+function(enable_ccache)
+    set(options QUIET)
+    set(oneValueArgs)
+    set(multiValueArgs)
+    cmake_parse_arguments(CCACHE "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )
+
+    find_program(CCACHE_BIN ccache REQUIRED)
+    if(NOT CCACHE_BIN)
+        message(STATUS "Morpheus: CCache not found. Rebuilding all the source files.")
+        message(DEBUG "Morpheus: If CCache compiler cache is desired then ensure this is a supported platform and ensure the Conan buildenv is active")
+        return()
+    endif()
+
+    message(STATUS "Morpheus Generator: ${CMAKE_GENERATOR}") #debug
+    message(STATUS "Morpheus: CCache found: ${CCACHE_BIN}. Enabling ccache as compiler cache tool.")
+    if(${CMAKE_GENERATOR} MATCHES "Ninja|Makefiles")
+        message(STATUS "found generator Ninja|Makefiles") #debug
+        foreach(lang IN ITEMS ${MORPHEUS_LANGUAGES})
+            set(CMAKE_${lang}_COMPILER_LAUNCHER ${CCACHE_BIN})
```

Comment:
This is a good starting point, but this is where it gets more interesting.  That is because we need to be able to pass through settings to `ccache`.  The mechanism for this in `ccache` is to set environment variables for each parameter that is controllable. 

CMake has a very specific way of doing this, you can use the `${CMAKE_COMMAND}` variable to invoke a CMake subprocess to launch your process with environment variable propagated: https://cmake.org/cmake/help/latest/manual/cmake.1.html#run-a-command-line-tool.   So for example this could be achieved using the `${CMAKE_COMMAND} -E env ${ccacheEnvVariables} ${CCACHE_BIN}` command passed to the compiler launcher.

To quote the CMake Professional book's section on Ccache:

```For most projects, the above example is too simplistic. It does not set any Ccache environment variables, it relies purely on the user-level configuration of that tool. CMake’s command mode can be used to set the relevant environment variables when invoking Ccache on any platform. The following is a more realistic example: 

find_program(CCACHE_EXECUTABLE ccache)
if(CCACHE_EXECUTABLE)
   set(ccacheEnv CCACHE_SLOPPINESS=pch_defines,time_macros) # NOTE: Ccache 4.2+ required for reliable CUDA support
   foreach(lang IN ITEMS C CXX OBJC OBJCXX CUDA)
        set(CMAKE_${lang}_COMPILER_LAUNCHER ${CMAKE_COMMAND} -E env ${ccacheEnv} ${CCACHE_EXECUTABLE}      )
   endforeach()
endif()```

Reply:
Anyway, in summary the next step is to look at the Ccache documentation and see which variable we should be setting and then expose them by CMake variable or parameter to the `enable_cache` method.

Reply:
Good approach. I will try to start with the following that seem useful:
```
CCACHE_SLOPPINESS=pch_defines,time_macros
CCACHE_PREFIX=icecc (if Icecream executable is present)
CCACHE_LOGFILE=~/.ccache/ccache_logfile.txt (to understand in the detail what ccache is doing)
```
I guess we will discover the others that we need once we start to build with ccache consistently. What do you think?

Reply:
Yes, completely agree!  These variable sounds like a good start. I need to spend time time reading through the docs to understand all of the options but we can add incrementally to it as we build our understanding.